### ISSUE TITLE
Make sure ServeMux is used in the right order

### ISSUE THREAD

**Issue by empijei** [google/go-safeweb#117]
The `ServeMux` construction goes through phases:

1. Plugin registration
2. Handlers registration and configuration
3. Execution

Nothing that is done in one of the phases should be done in one of the others.

I think we should enforce this somehow, by either providing builders for the phases or panicking if something is done out of order at startup, and then ignoring what happens at runtime (e.g. registering a plugin or a handler after a request has been served will log an error and do nothing).

**Comment by bsiegert** [comment]
In Goa, we have state variables that we set and check with atomics and return an error if in the wrong phase. For instance, if you want to add a module to a running server, it will be rejected this way.

**Comment by empijei** [comment]
> return an error if in the wrong phase

Wouldn't it be better to try and avoid this at compile time or startup time?

**Comment by empijei** [comment]
My proposal is an API that builds the mux in stages.

For example (this is just on the top of my head, don't take it as a proposal):
```go
myHandler := safehttp.ServeMuxBuilder(dispatcher, domains)
  .InstallInterceptor(...)
  .InstallInterceptor(...)
  .Mux()
  .Handle(...)
  .Handle(...)
  .Handler()
```

Where:
* `ServeMuxBuilder` returns something that only has the `InstallInterceptor` and `Mux` methods.
* `Mux` returns something that only has the `Handle` and `Handler` methods
* `Handler` returns a `safehttp.Handler`

**Comment by empijei** [comment]
@kele WDYT?

**Comment by kele** [comment]
I like it.

**Comment by empijei** [comment]
Alternatively we could just error or make out-of-order operations do nothing and log the error.

We could also panic.

I would just like our code to not have to care about things happening in an unreasonable sequence.

**Comment by maramihali** [comment]
Should this be v0?

**Comment by kele** [comment]
No. Focus on existing V0 tasks.

**Comment by kele** [comment]
Isn't this fixed by the `ServeMuxBuilder` from #191?
